### DISCUSSION TITLE
Jump to the beginning of the stream

### DISCUSSION THREAD

**Discussion by andreiltd** [winnow-rs/winnow#593] [Q&A] [answered]
Hello,

I'm having a hard time figuring out how to jump to the beginning of the stream.

It's easy to do with a `Checkpoint`, but only if the checkpoint is in the same scope as reset(). However, I need to reset the stream to a position relative to the beginning of the stream, somewhere deeper in the parsing process. I could include an initial checkpoint in the `Stateful` stream and use that, but it doesn't feel very elegant.

My stream would look like this (I need location tracking):

```rust
pub(crate) type Input<'a> = Stateful<Located<&'a [u8]>, State<'a>>;

pub(crate) struct State<'a> {
    pub start: Option<Checkpoint<Checkpoint<Checkpoint<&'a [u8], &'a [u8]>, Located<&'a [u8]>>, Stateful<Located<&'a [u8]>, State<'_>>>>
}
```

Is this the only way? Am I missing something obvious? Thanks!

**Comment by epage** [comment]
There is not a built-in way of doing this.

I've not seen this yet in another parsing library and don't have any use cases for it myself.  To evaluate whether this is worth supporting natively, we'd need to better understand the use case.  Technically, it wouldn't be too hard for a `Located` to support this as it knows the original input value.

Another good way to evaluate whether to support this natively is to provide it in a third-party solution and see what the adoption looks like.  There is nothing special to the design of custom stream types; they can be implemented outside of `winnow`.

**Reply by andreiltd** [reply]
Hey, thanks for taking the time to look into this,

My use case involves parsing an HDF5 file that contains addresses relative to the beginning of the file. During parsing, I need to jump to these addresses, parse the data, and then return to the original location before the jump.

For example, B-Tree data in HDF5 file contains addresses of sibling and child nodes:
https://docs.hdfgroup.org/hdf5/v1_14/_f_m_t3.html#title2

 What I need in this case is an equivalent of `fseek(stream, pointer, SEEK_SET)` although I understand it is not a very common use case.

**Reply by epage** [reply, marked as answer]
Oh interesting!  I wonder how other higher level parsers (ie not using `fseek`) deal with this.

For now, I've added `Located::reset_to_start` in #595 which is available in 0.6.19
